### PR TITLE
combined all DWR calls into one for withdrawal

### PR DESCRIPTION
**Pull request by Sidu28**
Combines calls to DWR and EPM into one call across multiple withdrawals.  Before, we were making calls for every withdrawal that was proven

### CODE REVIEW

File: src/contracts/pods/EigenPod.sol [hunk 1]
```diff
@@ -317,8 +317,18 @@ contract EigenPod is IEigenPod, Initializable, ReentrancyGuardUpgradeable, Eigen
             (withdrawalProofs.length == withdrawalFields.length), "EigenPod.verifyAndProcessWithdrawals: inputs must be same length"
         );
 
+        uint256 totalWithdrawalAmount;
```
Comment (gpsanant) [inline src/contracts/pods/EigenPod.sol:320]: totalAmountToSend

File: src/contracts/pods/EigenPod.sol [hunk 2]
```diff
@@ -546,8 +557,7 @@ contract EigenPod is IEigenPod, Initializable, ReentrancyGuardUpgradeable, Eigen
          * If the validator status is inactive, then withdrawal credentials were never verified for the validator,
          * and thus we cannot know that the validator is related to this EigenPod at all!
          */
-        ValidatorInfo memory validatorInfo = _validatorPubkeyHashToInfo[validatorPubkeyHash];
```
Comment (gpsanant) [inline src/contracts/pods/EigenPod.sol:549]: why are we loading twice? stack too deep?

File: src/contracts/pods/EigenPod.sol [hunk 3]
```diff
@@ -605,8 +614,8 @@ contract EigenPod is IEigenPod, Initializable, ReentrancyGuardUpgradeable, Eigen
         address recipient,
         uint64 withdrawalAmountGwei,
         ValidatorInfo memory validatorInfo
-    ) internal {
-        uint256 amountToSend;
+    ) internal returns(WithdrawalInfo memory) {
```
Comment (gpsanant) [inline src/contracts/pods/EigenPod.sol:617]: imo the struct is bulky and not needed.

returning uint256,int256 is better

Comment (Sidu28) [reply src/contracts/pods/EigenPod.sol:617]: stack too deep ma boi

File: src/contracts/pods/EigenPod.sol (hunk 2 above)
Comment (Sidu28) [reply src/contracts/pods/EigenPod.sol:549]: yup

File: src/contracts/pods/EigenPod.sol (hunk 1 above)
Comment (Sidu28) [reply src/contracts/pods/EigenPod.sol:320]: fixed

File: src/contracts/pods/EigenPod.sol [hunk 4]
```diff
@@ -317,8 +317,18 @@ contract EigenPod is IEigenPod, Initializable, ReentrancyGuardUpgradeable, Eigen
...
+        uint256 totalAmountToSend;
+        int256 totalSharesDelta;
         for (uint256 i = 0; i < withdrawalFields.length; i++) {
-            _verifyAndProcessWithdrawal(oracleTimestamp, withdrawalProofs[i], validatorFieldsProofs[i], validatorFields[i], withdrawalFields[i]);
+            WithdrawalInfo memory withdrawalInfo = _verifyAndProcessWithdrawal(oracleTimestamp, withdrawalProofs[i], validatorFieldsProofs[i], validatorFields[i], withdrawalFields[i]);
+            totalAmountToSend += withdrawalInfo.amountToSend;
+            totalSharesDelta += withdrawalInfo.sharesDelta;
+        }
+        // send ETH to the `recipient` via the DelayedWithdrawalRouter, if applicable
+        _sendETH_AsDelayedWithdrawal(podOwner, totalAmountToSend);
```
[the remark this answers is not in the document]
Comment (Sidu28) [reply src/contracts/pods/EigenPod.sol:328]: fixed

File: src/contracts/interfaces/IEigenPod.sol [hunk 5]
```diff
@@ -50,6 +50,13 @@ interface IEigenPod {
         VALIDATOR_STATUS status;
     }
 
+    struct WithdrawalInfo {
+        //amount to send to a podOwner from a proven withdrawal
+        uint256 amountToSend;
+        //difference in shares to be recorded in the eigenPodManager
+        int256 sharesDelta;
+    }
```
Comment (ChaoticWalrus) [inline src/contracts/interfaces/IEigenPod.sol:58]: is there something else we can name this, perhaps more specific or descriptive?
even just adding a top-level comment to the struct definition (perhaps just capturing its usage) might help.
also going to note a nit about spacing; should have a space between the slashes and start of comment

File: src/contracts/pods/EigenPod.sol [hunk 6]
```diff
@@ -317,8 +317,20 @@ contract EigenPod is IEigenPod, Initializable, ReentrancyGuardUpgradeable, Eigen
...
+            WithdrawalInfo memory withdrawalInfo = _verifyAndProcessWithdrawal(oracleTimestamp, withdrawalProofs[i], validatorFieldsProofs[i], validatorFields[i], withdrawalFields[i]);
+            totalAmountToSend += withdrawalInfo.amountToSend;
+            totalSharesDelta += withdrawalInfo.sharesDelta;
+        }
+        // send ETH to the `recipient` via the DelayedWithdrawalRouter, if applicable
+        if (totalAmountToSend != 0) {
+            _sendETH_AsDelayedWithdrawal(podOwner, totalAmountToSend);
+        }
+        //update podOwner's shares in the strategy manager
+        if (totalSharesDelta != 0){
```
Comment (ChaoticWalrus) [inline src/contracts/pods/EigenPod.sol:332]: nit: spacing
also noticing that we are inputting the `oracleTimestamp` into `_verifyAndProcessWithdrawal` repeatedly -- can we not move that logic out of the loop here too?

File: src/contracts/interfaces/IEigenPod.sol (hunk 5 above)
Comment (ChaoticWalrus) [reply src/contracts/interfaces/IEigenPod.sol:58]: suggestion:
```
    /** 
     * @notice struct used to store amounts related to proven withdrawals in memory. Used to help
     * manage stack depth and optimize the number of external calls, when batching withdrawal operations.
     */
    struct ProvenWithdrawal {
        // amount to send to a podOwner from a proven withdrawal
        uint256 amountToSend;
        // difference in shares to be recorded in the eigenPodManager, as a result of the withdrawal
        int256 sharesDelta;
    }
```

**Review by ChaoticWalrus — CHANGES_REQUESTED** [review]
This PR is a really good idea. Added a few comments that I'd like to see addressed.

File: src/contracts/interfaces/IEigenPod.sol (hunk 5 above)
Comment (Sidu28) [reply src/contracts/interfaces/IEigenPod.sol:58]: fixed

File: src/contracts/pods/EigenPod.sol [hunk 7]
```diff
@@ -583,17 +595,16 @@ contract EigenPod is IEigenPod, Initializable, ReentrancyGuardUpgradeable, Eigen
...
 
             /**
             * if the validator's withdrawable epoch is less than or equal to the slot's epoch, then the validator has fully withdrawn because
             * a full withdrawal is only processable after the withdrawable epoch has passed.
             */
             // reference: uint64 withdrawableEpoch = Endian.fromLittleEndianUint64(validatorFields[BeaconChainProofs.VALIDATOR_WITHDRAWABLE_EPOCH_INDEX]);
-            if (Endian.fromLittleEndianUint64(validatorFields[BeaconChainProofs.VALIDATOR_WITHDRAWABLE_EPOCH_INDEX]) <= slot/BeaconChainProofs.SLOTS_PER_EPOCH) {
-                _processFullWithdrawal(validatorIndex, validatorPubkeyHash, withdrawalHappenedTimestamp, podOwner, withdrawalAmountGwei, validatorInfo);
+            if (Endian.fromLittleEndianUint64(validatorFields[BeaconChainProofs.VALIDATOR_WITHDRAWABLE_EPOCH_INDEX]) <= (Endian.fromLittleEndianUint64(withdrawalProof.slotRoot))/BeaconChainProofs.SLOTS_PER_EPOCH) {
+                return _processFullWithdrawal(validatorIndex, validatorPubkeyHash, withdrawalHappenedTimestamp, podOwner, withdrawalAmountGwei, _validatorPubkeyHashToInfo[validatorPubkeyHash]);
```
[the remark this answers is not in the document]
Comment (Sidu28) [reply src/contracts/pods/EigenPod.sol:605]: fixed, let me know if the break up follows our general style guideline

File: src/contracts/pods/EigenPod.sol (hunk 6 above)
Comment (Sidu28) [reply src/contracts/pods/EigenPod.sol:332]: good call, fixed

File: src/contracts/pods/EigenPod.sol [hunk 8]
```diff
@@ -537,6 +550,7 @@ contract EigenPod is IEigenPod, Initializable, ReentrancyGuardUpgradeable, Eigen
          * *prior* to the proof provided in the `verifyWithdrawalCredentials` function.
          */
         proofIsForValidTimestamp(Endian.fromLittleEndianUint64(withdrawalProof.timestampRoot))
+        returns (VerifiedWithdrawal memory)
     {
         uint64 withdrawalHappenedTimestamp = _computeTimestampAtSlot(Endian.fromLittleEndianUint64(withdrawalProof.slotRoot));
```
Comment (ChaoticWalrus) [inline src/contracts/pods/EigenPod.sol:555]: does this timestamp check still make sense here / is it even correct?
seems quite odd that we essentially have 3 different timestamps here -- the `oracleTimestamp`, `withdrawalProof.timestampRoot`, and converting the `withdrawalProof.slotRoot`  to a timestamp !

Comment (Sidu28) [reply src/contracts/pods/EigenPod.sol:557]: G added timestamp root as a field in the proof but this line didn't get updated - thanks for noting it, fixing it now

Comment (ChaoticWalrus) [reply src/contracts/pods/EigenPod.sol]: OK excellent, really glad to see the simplification and adoption of more consistency here.
Should we move the timestamp check (`proofIsForValidTimestamp(Endian.fromLittleEndianUint64(withdrawalProof.timestampRoot))`) outside of the loop, so we aren't checking it each time?

Comment (ChaoticWalrus) [reply src/contracts/pods/EigenPod.sol]: nevermind, seems like we have to do this in the loop, am re-resolving this thread.

**Review by ChaoticWalrus — APPROVED** [review]

File: src/contracts/pods/EigenPod.sol (hunk 8 above)
Comment (Sidu28) [reply src/contracts/pods/EigenPod.sol]: Discussed on slack, this is the withdrawal timestamp not the oracle timestamp, it needs to be checked for each proof

**Pull request merged**